tests/test_numpy_dtypes.cpp

```diff
@@ -404,10 +404,10 @@ TEST_SUBMODULE(numpy_dtypes, m) {
         return l;
     });
 
-    // test_dtype
+    // test_dtype (ulong was missing, but uint default changed on windows so use L)
     std::vector<const char *> dtype_names{
-        "byte",    "short",   "intc",        "int_",  "longlong",   "ubyte",       "ushort",
-        "uintc",   "uint",    "ulonglong",   "half",  "single",     "double",      "longdouble",
+        "byte",    "short",   "intc",        "long",  "longlong",   "ubyte",       "ushort",
+        "uintc",   "L",       "ulonglong",   "half",  "single",     "double",      "longdouble",
         "csingle", "cdouble", "clongdouble", "bool_", "datetime64", "timedelta64", "object_"};
```

Comment:
This removes two dtype names from the list (and adds two).

From my naive numpy outsider perspective I'm getting worried about things being removed from tests. So I tried the diff below and it passes (on the one platform I tried: g++ (Debian 13.2.0-10) 13.2.0, Python 3.11.7).

Basically, previously we promised that `int_` and `uint` have a certain behavior, but now we're not promising that anymore. To be on the safe side, would it make sense to maintain testing with those dtype names?

```diff
diff --git a/tests/test_numpy_dtypes.cpp b/tests/test_numpy_dtypes.cpp
index 2141d272..0347fe9e 100644
--- a/tests/test_numpy_dtypes.cpp
+++ b/tests/test_numpy_dtypes.cpp
@@ -408,7 +408,8 @@ TEST_SUBMODULE(numpy_dtypes, m) {
     std::vector<const char *> dtype_names{
         "byte",    "short",   "intc",        "long",  "longlong",   "ubyte",       "ushort",
         "uintc",   "L",       "ulonglong",   "half",  "single",     "double",      "longdouble",
-        "csingle", "cdouble", "clongdouble", "bool_", "datetime64", "timedelta64", "object_"};
+        "csingle", "cdouble", "clongdouble", "bool_", "datetime64", "timedelta64", "object_",
+        "int_", "uint"};

     m.def("print_dtypes", []() {
         py::list l;
diff --git a/tests/test_numpy_dtypes.py b/tests/test_numpy_dtypes.py
index 402b940a..075f827e 100644
--- a/tests/test_numpy_dtypes.py
+++ b/tests/test_numpy_dtypes.py
@@ -173,8 +173,8 @@ def test_dtype(simple_dtype):
         np.zeros(1, m.trailing_padding_dtype())
     )

-    expected_chars = "bhilqBHILQefdgFDG?MmO"
-    assert m.test_dtype_kind() == list("iiiiiuuuuuffffcccbMmO")
+    expected_chars = "bhilqBHILQefdgFDG?MmOlL"
+    assert m.test_dtype_kind() == list("iiiiiuuuuuffffcccbMmOiu")
     assert m.test_dtype_char_() == list(expected_chars)
     assert m.test_dtype_num() == [np.dtype(ch).num for ch in expected_chars]
     assert m.test_dtype_byteorder() == [np.dtype(ch).byteorder for ch in expected_chars]
```

Reply:
My understanding is that `py::dtype("string")` is identical to `np.dtype("string")`, so I wasn't too worried about reducing the test a tiny bit.

`L` (single letter) was just the only way to spell `ulong` that works on NumPy 2 and quite old NumPy versions unfortunately if you wish to spell `unsigned long`.  (Thus the other comment)

Your diff will fail tests on windows, thus the change.  So you could do it, but you would have to use `np.dtype("int_").char` to compare it with.

Reply:
> Your diff will fail tests on windows, thus the change.

That's exactly what I mean by "something is getting swept under the rug here."

What is that failure?

Is it recoverable?

If yes: we should add that as a test here, and point here from the changelog.

If no (e.g. segfault): we need to document that and have a **bold warning** in the changelog.

Reply:
Well, I have added more tests and put a suggestion in the suggested change-log.  We clearly disagree on what you need to document in pybind11 vs. just referring to the NumPy docs so I am sorry, but I don't know what "bold warning" you would hope to see, I added this change as an example for what NumPy 2 breaks (it is possible a more relevant one, but certainly not the only one).